graphql/src/execution/execution.rs

```diff
@@ -306,49 +353,64 @@ pub fn execute_root_selection_set(
+    };
     let result = if let Some(key) = key {
-        let cached = QUERY_HERD_CACHE.cached_query(key, || {
-            execute_root_selection_set_uncached(ctx, selection_set, root_type)
-        });
-        MaybeCached::Cached(cached)
+        QUERY_HERD_CACHE.cached_query(key, run_query)
     } else {
-        let not_cached = execute_root_selection_set_uncached(ctx, selection_set, root_type);
-        MaybeCached::NotCached(not_cached)
+        run_query()
     };
+    if herd_hit {
+        ctx.cache_status.store(CacheStatus::Shared);
+    }
 
     // Check if this query should be cached.
-    if let (MaybeCached::Cached(cached), Some(key), Some(block_ptr)) = (&result, key, block_ptr) {
-        // Share errors from the herd cache, but don't store them in generational cache.
-        // In particular, there is a problem where asking for a block pointer beyond the chain
-        // head can cause the legitimate cache to be thrown out.
-        if cached.is_ok() {
-            let mut cache = QUERY_CACHE.write().unwrap();
-
-            // If there is already a cache by the block of this query, just add it there.
-            if let Some(cache_by_block) = cache.iter_mut().find(|c| c.block == block_ptr) {
-                cache_by_block.cache.insert(key, cached.cheap_clone());
-            } else if *QUERY_CACHE_BLOCKS > 0 {
-                // We're creating a new `CacheByBlock` if:
-                // - There are none yet, this is the first query being cached, or
-                // - `block_ptr` is of higher or equal number than the most recent block in the cache.
-                // Otherwise this is a historical query which will not be cached.
-                let should_insert = match cache.iter().next() {
-                    None => true,
-                    Some(highest) if highest.block.number <= block_ptr.number => true,
-                    Some(_) => false,
-                };
-
-                if should_insert {
-                    if cache.len() == *QUERY_CACHE_BLOCKS {
-                        // At capacity, so pop the oldest block.
-                        cache.pop_back();
-                    }
+    // Share errors from the herd cache, but don't store them in generational cache.
+    // In particular, there is a problem where asking for a block pointer beyond the chain
+    // head can cause the legitimate cache to be thrown out.
+    // It would be redundant to insert herd cache hits.
+    let no_cache = herd_hit || result.has_errors();
+    if let (false, Some(key), Some(block_ptr)) = (no_cache, key, block_ptr) {
+        let mut cache = QUERY_CACHE.write().unwrap();
+
+        // If there is already a cache by the block of this query, just add it there.
+        if let Some(cache_by_block) = cache.iter_mut().find(|c| c.block == block_ptr) {
+            let cache_insert = cache_by_block.insert(key, result.cheap_clone());
+            if cache_insert {
+                ctx.cache_status.store(CacheStatus::Insert);
```

Comment:
If the status is `Shared` then `herd_hit` is true, therefore `no_cache` is true, so we won't even try to insert in the query cache and this line won't be executed. So that case should always be `Shared`, `Insert`.